pkg/tfbridge/detailed_diff.go

```diff
@@ -0,0 +1,484 @@
+	case pulumirpc.PropertyDiff_DELETE:
+		return &pulumirpc.PropertyDiff{Kind: pulumirpc.PropertyDiff_DELETE_REPLACE}
+	case pulumirpc.PropertyDiff_UPDATE:
+		return &pulumirpc.PropertyDiff{Kind: pulumirpc.PropertyDiff_UPDATE_REPLACE}
+	default:
+		return diff
+	}
+}
+
+func isForceNew(etf shim.Schema, eps *SchemaInfo) bool {
+	if etf != nil && etf.ForceNew() {
+		return true
+	}
+	if eps != nil && eps.ForceNew != nil && *eps.ForceNew {
+		return true
+	}
+	return false
+}
+
+func mapHasReplacements(m map[string]*pulumirpc.PropertyDiff) bool {
+	for _, diff := range m {
+		if diff.GetKind() == pulumirpc.PropertyDiff_ADD_REPLACE ||
+			diff.GetKind() == pulumirpc.PropertyDiff_DELETE_REPLACE ||
+			diff.GetKind() == pulumirpc.PropertyDiff_UPDATE_REPLACE {
+			return true
+		}
+	}
+	return false
+}
+
+func simplifyDiff(
+	diff map[string]*pulumirpc.PropertyDiff, key resource.PropertyKey, old, new resource.PropertyValue,
+	oldOk, newOk bool, etf interface{}, eps *SchemaInfo,
+) (map[string]*pulumirpc.PropertyDiff, error) {
+	baseDiff := makeBaseDiff(old, new, oldOk, newOk)
+	if baseDiff != Undecided {
+		etf, ok := etf.(shim.Schema)
+		if !ok {
+			etf = nil
+		}
+		propDiff := baseDiffToPropertyDiff(baseDiff, etf, eps)
+		if propDiff == nil {
+			return nil, nil
+		}
+		if mapHasReplacements(diff) {
+			propDiff = promoteToReplace(propDiff)
+		}
+		return map[string]*pulumirpc.PropertyDiff{string(key): propDiff}, nil
+	}
+	return nil, errors.New("diff is not simplified")
+}
+
+func propertyDiffResult(etf shim.Schema, eps *SchemaInfo, diff *pulumirpc.PropertyDiff) *pulumirpc.PropertyDiff {
+	// See pkg/cross-tests/diff_cross_test.go
+	// TestAttributeCollectionForceNew, TestBlockCollectionForceNew, TestBlockCollectionElementForceNew
+	// for a full case study of replacements in TF
+	if isForceNew(etf, eps) {
+		return promoteToReplace(diff)
+	}
+	return diff
+}
+
+func makeTopPropDiff(
+	old, new resource.PropertyValue,
+	oldOk, newOk bool,
+	etf shim.Schema,
+	eps *SchemaInfo,
+) *pulumirpc.PropertyDiff {
+	baseDiff := makeBaseDiff(old, new, oldOk, newOk)
+	if baseDiff != Undecided {
+		return baseDiffToPropertyDiff(baseDiff, etf, eps)
+	}
+
+	if !old.DeepEquals(new) {
+		return propertyDiffResult(etf, eps, &pulumirpc.PropertyDiff{Kind: pulumirpc.PropertyDiff_UPDATE})
+	}
+	return nil
+}
+
+func makePropDiff(
+	ctx context.Context,
+	key resource.PropertyKey,
+	etf shim.Schema,
+	eps *SchemaInfo,
+	old, new resource.PropertyValue,
+	oldOk, newOk bool,
+) map[string]*pulumirpc.PropertyDiff {
+	// ignore dunder properties - these are internal to Pulumi and should not be surfaced in the diff
+	if isDunder(key) {
+		return nil
+	}
+	res := make(map[string]*pulumirpc.PropertyDiff)
+
+	if etf == nil {
+		// If the schema is nil, we just return the top-level diff
+		topDiff := makeTopPropDiff(old, new, oldOk, newOk, etf, eps)
+		if topDiff == nil {
+			return nil
+		}
+		res[string(key)] = topDiff
+		return res
+	}
+
+	if isFlattened(etf, eps) {
```

Comment:
This is really rough to read :) We're missing some facilities perhaps can check something in quickly for this PR to build on? Ideally the case analysis reads very very simply. We're comparing two PropertyValues. What schema typing information we have available?

```
Case 1: object type with SchemaMap and map[string]*SchemaInfo
Case 2: map type with element Schema + SchemaInfo
Case 3: list type with element Schema + SchemaInfo
Case 4: set type with element Schema + SchemaInfo
Case 5: no type information, lost typing
Case 6: etc type that is not a collections (scalar types)
```

If we could make the case analysis read just like that, it'd be cool. Shim layer should make this easy. I wonder again if with  LookupSchemaPath we're almost there but probably not quite - I didn't have quality review on that code yet and it might need a few touches.